executor/joiner.go

```diff
@@ -176,108 +181,111 @@ type semiJoiner struct {
 }
 
-func (j *semiJoiner) onMissMatch(outer chunk.Row, chk *chunk.Chunk) {
+func (j *semiJoiner) onMissMatch(hasNull bool, outer chunk.Row, chk *chunk.Chunk) {
 }
 
 type antiSemiJoiner struct {
 	baseJoiner
 }
 
 // tryToMatch implements joiner interface.
-func (j *antiSemiJoiner) tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) (matched bool, err error) {
+func (j *antiSemiJoiner) tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) (matched bool, hasNull bool, err error) {
 	if inners.Len() == 0 {
-		return false, nil
+		return false, false, nil
 	}
 
 	if len(j.conditions) == 0 {
 		inners.ReachEnd()
-		return true, nil
+		return true, false, nil
 	}
 
 	for inner := inners.Current(); inner != inners.End(); inner = inners.Next() {
 		j.makeShallowJoinRow(j.outerIsRight, inner, outer)
 
-		matched, err = expression.EvalBool(j.ctx, j.conditions, j.shallowRow.ToRow())
+		matched, isNull, err := expression.EvalBool(j.ctx, j.conditions, j.shallowRow.ToRow())
 		if err != nil {
-			return false, errors.Trace(err)
+			return false, false, errors.Trace(err)
 		}
 		if matched {
 			inners.ReachEnd()
-			return true, nil
+			return true, false, nil
 		}
+		hasNull = hasNull || isNull
 	}
-	return false, nil
+	return false, hasNull, nil
 }
 
-func (j *antiSemiJoiner) onMissMatch(outer chunk.Row, chk *chunk.Chunk) {
-	chk.AppendRow(outer)
+func (j *antiSemiJoiner) onMissMatch(hasNull bool, outer chunk.Row, chk *chunk.Chunk) {
+	if !hasNull {
+		chk.AppendRow(outer)
+	}
 }
 
 type leftOuterSemiJoiner struct {
 	baseJoiner
 }
 
 // tryToMatch implements joiner interface.
-func (j *leftOuterSemiJoiner) tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) (matched bool, err error) {
+func (j *leftOuterSemiJoiner) tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) (matched bool, hasNull bool, err error) {
 	if inners.Len() == 0 {
-		return false, nil
+		return false, false, nil
 	}
 
 	if len(j.conditions) == 0 {
 		j.onMatch(outer, chk)
 		inners.ReachEnd()
-		return true, nil
+		return true, false, nil
 	}
 
 	for inner := inners.Current(); inner != inners.End(); inner = inners.Next() {
 		j.makeShallowJoinRow(false, inner, outer)
 
-		matched, err = expression.EvalBool(j.ctx, j.conditions, j.shallowRow.ToRow())
+		matched, _, err = expression.EvalBool(j.ctx, j.conditions, j.shallowRow.ToRow())
 		if err != nil {
-			return false, errors.Trace(err)
+			return false, false, errors.Trace(err)
 		}
 		if matched {
 			j.onMatch(outer, chk)
 			inners.ReachEnd()
-			return true, nil
+			return true, false, nil
 		}
 	}
-	return false, nil
+	return false, false, nil
 }
 
 func (j *leftOuterSemiJoiner) onMatch(outer chunk.Row, chk *chunk.Chunk) {
 	chk.AppendPartialRow(0, outer)
 	chk.AppendInt64(outer.Len(), 1)
 }
 
-func (j *leftOuterSemiJoiner) onMissMatch(outer chunk.Row, chk *chunk.Chunk) {
+func (j *leftOuterSemiJoiner) onMissMatch(hasNull bool, outer chunk.Row, chk *chunk.Chunk) {
```

Comment:
After maintaining the correct `hasNull` value, if `hasNull` is true, we should append a NULL value to make the correct join result.

Reply:
Ditto